### DISCUSSION TITLE
Merkle Tree Generator

### DISCUSSION THREAD

**Discussion by graemecode** [wevm/viem#529] [Idea / Feature Request]
## Summary

_Merkle Tree_ proofs are a crucial primitive in the web3 ecosystem. I believe they are still vastly underused by general web3 applications, given that they allow proving specific data in a concise way – a core goal in web3 – without needing full rollups.

I've been using Merkle Trees in a few applications, starting with Mirror's [Splits](https://dev.mirror.xyz/V_7Jp1hy_g8bz-J1B4Wb5KYSmj5Lt4W7q7cw0noxJsU) and now in some aspects of a new project.

By providing a simple open-source standard for generating Merkle Trees, we could be helpful in popularizing their use in more applications.

## Basic Example

Here's an example of how we could use an interface from Viem to generate a proof of membership feature:

### Tree Generation Interface

```typescript
import { createMerkleTree } from "viem";

const data = [
 { 
   owner: "0xCC65fA278B917042822538c44ba10AD646824026",
   handle: "kanye"
  },
 {
   owner: "0xFfFe0eCAb95CdCd77453DAf5454d4d57Ca283838",
   handle: "mazzy"
 }
];

const merkleTree = createMerkleTree(data);

const { root: merkleRoot } = merkleTree;
```

The merkleRoot hash can now be stored on an L2 for a small amount of gas, and used to prove 
that `0xCC65fA278B917042822538c44ba10AD646824026` owns the handle "kayne" for some application 
(without that application needing to run its full signup process on an L2.)

### Verification Interface

```typescript
import { verifyMerkleProof } from "viem";

// Get the proof for kayne's handle:
const [kayneProof] = tree.proofs;

// Get kayne's data
const [kayne] = data;

const ownsKayne = verifyMerkleProof(kayne, kayneProof);
// => true
```

## Discussion

I'm curious if this is something that is aligned with Viem, and if so I'd be glad to contribute code/review/feedback.

Thanks
– Graeme

**Comment by jxom** [comment]
Sounds interesting! What prior art is out there right now in regards to generating/verifying merkle trees with JavaScript?

**Reply by graemecode** [reply]
I've just written my own so far. It's not too complicated, but it would be good to have a community standard in the ecosystem!

**Comment by itsmnthn** [comment]
Yeah, this would be gud to have 😍, Also using custom thing

Don't know what the standard is but I'm using it like this to send the proof to contract call

```TypeScript
import keccak256 from 'keccak256'
import MerkleTree from 'merkletreejs'


/**
 *  Data can be standard format
 * { owner: Address, leave: Address }[]
 * { [key in Address]: Address } // owner as key and leave as value
 */
const leaves = [ /** leaves array or derived array from json */]

const merkleTree = new MerkleTree(leaves, keccak256, { sortPairs: true })
const proof = merkleTree.getHexProof(leave)

walletClient.writeContract({
  // ...
  
  args: [proof, /** .... */]
  
  // ...
})

```

**Comment by aalmada** [comment]
I'm using [MerkleTree.js](https://github.com/merkletreejs/merkletreejs) with Viem. For example, to generate a tree with leafs that contain an address and an id, I use the following:

```typescript
import { keccak256, encodePacked, Hex } from 'viem';
import { MerkleTree } from 'merkletreejs';

type MerkleLeaf = { account: Hex; id: bigint };

const getHash = (leaf: MerkleLeaf): Hex => keccak256(encodePacked(['uint256', 'address'], [leaf.id, leaf.account]));

const createMerkleTree = (leaves: MerkleLeaf[]): MerkleTree =>
new MerkleTree(leaves.map(getHash), keccak256, { sort: true });
```

You can see it working [here](https://github.com/aalmada/honest-nft/blob/1751352884608cfeadfbc27ec23101756a26b8e8/packages/blockchain/test/MyNft.ts#L81).

In case you plan to implement a MerkleTree type, I've found this interesting article: ["A Hackers Guide to Layer 2: Zero Merkle Trees from Scratch"](https://medium.com/@carterfeldman/a-hackers-guide-to-layer-2-zero-merkle-trees-from-scratch-d612ea846016).
